{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## fastq2vcf\n",
    "### Tutorial intended to cover the analysis of NGS human data. In particular the processing of fastq files in order to get variants in VCF format. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fastq Downloading\n",
    "\n",
    "We are going to use public data from the Simons Diversity Project. Specifically 10 selected genomes from different populations. Here below a table with metadata downloaded from ENA repository."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>PRJEB9586</td>\n",
       "      <td>ERP010710</td>\n",
       "      <td>SAMEA3302753</td>\n",
       "      <td>ERS1042136</td>\n",
       "      <td>ERX1097974</td>\n",
       "      <td>ERR1019034</td>\n",
       "      <td>ERA486110</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/err/ERR101/004/ERR101...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/err/ERR101/004/ERR1019034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA3302753</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simons_data_sample_155</td>\n",
       "      <td>0</td>\n",
       "      <td>2015-09-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>PRJEB9586</td>\n",
       "      <td>ERP010710</td>\n",
       "      <td>SAMEA3302890</td>\n",
       "      <td>ERS1042137</td>\n",
       "      <td>ERX1097975</td>\n",
       "      <td>ERR1019035</td>\n",
       "      <td>ERA486110</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/err/ERR101/005/ERR101...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/err/ERR101/005/ERR1019035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA3302890</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simons_data_sample_224</td>\n",
       "      <td>0</td>\n",
       "      <td>2015-09-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>PRJEB9586</td>\n",
       "      <td>ERP010710</td>\n",
       "      <td>SAMEA3302855</td>\n",
       "      <td>ERS1042138</td>\n",
       "      <td>ERX1097976</td>\n",
       "      <td>ERR1019036</td>\n",
       "      <td>ERA486110</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/err/ERR101/006/ERR101...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/err/ERR101/006/ERR1019036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA3302855</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simons_data_sample_180</td>\n",
       "      <td>0</td>\n",
       "      <td>2015-09-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PRJEB9586</td>\n",
       "      <td>ERP010710</td>\n",
       "      <td>SAMEA3302840</td>\n",
       "      <td>ERS1042139</td>\n",
       "      <td>ERX1097977</td>\n",
       "      <td>ERR1019037</td>\n",
       "      <td>ERA486110</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA3302840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simons_data_sample_265</td>\n",
       "      <td>0</td>\n",
       "      <td>2015-09-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>PRJEB9586</td>\n",
       "      <td>ERP010710</td>\n",
       "      <td>SAMEA3302867</td>\n",
       "      <td>ERS1042140</td>\n",
       "      <td>ERX1097978</td>\n",
       "      <td>ERR1019038</td>\n",
       "      <td>ERA486110</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/err/ERR101/008/ERR101...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/err/ERR101/008/ERR1019038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA3302867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simons_data_sample_207</td>\n",
       "      <td>0</td>\n",
       "      <td>2015-09-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession secondary_study_accession sample_accession  \\\n",
       "0       PRJEB9586                 ERP010710     SAMEA3302753   \n",
       "1       PRJEB9586                 ERP010710     SAMEA3302890   \n",
       "2       PRJEB9586                 ERP010710     SAMEA3302855   \n",
       "3       PRJEB9586                 ERP010710     SAMEA3302840   \n",
       "4       PRJEB9586                 ERP010710     SAMEA3302867   \n",
       "\n",
       "  secondary_sample_accession experiment_accession run_accession  \\\n",
       "0                 ERS1042136           ERX1097974    ERR1019034   \n",
       "1                 ERS1042137           ERX1097975    ERR1019035   \n",
       "2                 ERS1042138           ERX1097976    ERR1019036   \n",
       "3                 ERS1042139           ERX1097977    ERR1019037   \n",
       "4                 ERS1042140           ERX1097978    ERR1019038   \n",
       "\n",
       "  submission_accession  tax_id scientific_name instrument_platform  ...  \\\n",
       "0            ERA486110    9606    Homo sapiens            ILLUMINA  ...   \n",
       "1            ERA486110    9606    Homo sapiens            ILLUMINA  ...   \n",
       "2            ERA486110    9606    Homo sapiens            ILLUMINA  ...   \n",
       "3            ERA486110    9606    Homo sapiens            ILLUMINA  ...   \n",
       "4            ERA486110    9606    Homo sapiens            ILLUMINA  ...   \n",
       "\n",
       "                                          sra_aspera  \\\n",
       "0  fasp.sra.ebi.ac.uk:/vol1/err/ERR101/004/ERR101...   \n",
       "1  fasp.sra.ebi.ac.uk:/vol1/err/ERR101/005/ERR101...   \n",
       "2  fasp.sra.ebi.ac.uk:/vol1/err/ERR101/006/ERR101...   \n",
       "3                                                NaN   \n",
       "4  fasp.sra.ebi.ac.uk:/vol1/err/ERR101/008/ERR101...   \n",
       "\n",
       "                                         sra_galaxy  cram_index_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/err/ERR101/004/ERR1019034             NaN   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/err/ERR101/005/ERR1019035             NaN   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/err/ERR101/006/ERR1019036             NaN   \n",
       "3                                               NaN             NaN   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/err/ERR101/008/ERR1019038             NaN   \n",
       "\n",
       "  cram_index_aspera cram_index_galaxy  sample_alias broker_name  \\\n",
       "0               NaN               NaN  SAMEA3302753         NaN   \n",
       "1               NaN               NaN  SAMEA3302890         NaN   \n",
       "2               NaN               NaN  SAMEA3302855         NaN   \n",
       "3               NaN               NaN  SAMEA3302840         NaN   \n",
       "4               NaN               NaN  SAMEA3302867         NaN   \n",
       "\n",
       "             sample_title  nominal_sdev first_created  \n",
       "0  simons_data_sample_155             0    2015-09-18  \n",
       "1  simons_data_sample_224             0    2015-09-18  \n",
       "2  simons_data_sample_180             0    2015-09-18  \n",
       "3  simons_data_sample_265             0    2015-09-18  \n",
       "4  simons_data_sample_207             0    2015-09-18  \n",
       "\n",
       "[5 rows x 51 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "projectID = 'PRJEB9586'\n",
    "! wget -O {projectID}.metadata 'http://www.ebi.ac.uk/ena/data/warehouse/filereport?accession=PRJEB9586&result=read_run&download=txt'\n",
    "\n",
    "file = pd.read_csv(projectID + '.metadata', sep='\\t')\n",
    "file.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EX2. Peer review the second article; submit it via web form "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EX3. Use the following [directed graph](https://courses.cs.ut.ee/MTAT.03.238/2019_fall/uploads/Main/directed_graph.png). Convert it into adjacency matrix representation and then convert the matrix into the graph where the original edges are replaced by \"3-hop\" edges or in other words - connect only vertices that are reachable in 3 steps of the original graph. \n",
    "\n",
    "![image 01](https://courses.cs.ut.ee/MTAT.03.238/2019_fall/uploads/Main/directed_graph.png \"Graph\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EX4. Make a transitive closure of the same graph - using three approaches. 1) G*G*G...; 2) making use of the following property - G2 = G*G, G4 = G2*G2, G8 = G4*G4 and 3) with Warshall algorithm. Verify that you got the same answer. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EX5. Implement a \"random walk\" procedure following links randomly (equal probabilities). Estimate the probability of being in any given node by varying how you deal with \"dead ends\" (e.g. node B) and \"nodes with no links into it\" (e.g. nodes D or A). First, try \"re-appearing anywhere\" from dead ends only. Secondly, when walking, with 20% probability jump to any random vertex in the network, and 80% of times select randomly and go to one of the neighbouring vertices. For both scenarios, provide the respective probabilities and identify the most important nodes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EX6 (Bonus 2p). Use the random walk data from the previous exercise. Implement the \"same\" random walk procedure but now clearly using matrix multiplications much like in finding transitive closures. Assign uneven probabilities to links - e.g. 80%-20% for two-link nodes (larger probability to alphabetically the first neighbour) and 60%-30%-10% for three link ones. Feel free to modify the input graph a bit. Introduce enough links so that you do not have \"dead ends\" or nodes where nothing leads into. E.g. I recommend introducing links G->H, B->K, K->H. But feel free to experiment. Of course, you should also figure out what happens to graph multiplications when you have not yet changed the graph. Your task is to report transition probabilities after N steps. How do you know when to stop multiplying? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EX7 (Bonus 2p). Extend your solution from EX6 and implement the Google Page Rank algorithm (here is a small [playground](https://learnforeverlearn.com/pagerank/)), the algorithm itself is very simple) where 20% of times you would re-appear anywhere in graph. Both using equal-probabilities links and the uneven 80-20, 60-30-10 probabilities. Report the outcome probabilities and compare to results in EX6. "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
